### ISSUE TITLE
Using NaN as fill value

### ISSUE THREAD

**Issue by gdkrmr** [constantinpape/z5#148]
I am writing the R wrapper and came across the following issue:
When specifying `NaN` as a fill value, it work fine in `z5::createDataset` and the resulting `z5::Dataset` object's `z5::Metadata` has `NaN` as a fill value. The issue is that the `.zarray` that gets written to disk saves `null` as a fill value and reading the dataset from disk fails.

Some things I figured out:
- The json library is behaving according to specs when saving `NaN` as `null`.
E.g. in the Firefox console:
```js
> console.log(JSON.stringify({"x": NaN}))
{"x":null}
```
- `Zarr.jl` simply saves `"NaN"` to disk.

Not sure what would be the best solution here. Read `null` as `NaN`, or save `NaN` as `"NaN"`, or both.

@meggart

**Comment by constantinpape** [comment]
@gdkrmr
Do you know what happens when saving a dataset with `NaN` fill-value with `zarr-python`?
What value does it save in the `.zarray`?

My preferred solution would be to do the same thing as zarr-python.

**Comment by gdkrmr** [comment]
No idea, I will investigate it.

**Comment by meggart** [comment]
See this related issue https://github.com/zarr-developers/zarr-python/issues/412

**Comment by gdkrmr** [comment]
```
import zarr
zarr.creation.create((9,9,9), fill_value = np.nan, store = "tmp.zarr")
```
gives `{..., "fill_value": "NaN", ...}`.

**Comment by meggart** [comment]
Yes for fill_value this is converted, but as stated in the issue, be careful when parsing .zattrs because there python will not put the NaN into strings.

**Comment by constantinpape** [comment]
@gdkrmr 
Ok, so I think we should save `"NaN"` to disc in this case as well.
The way to change this is changing [this line](https://github.com/constantinpape/z5/blob/master/include/z5/metadata.hxx#L83)
to something like
```
j["fill_value"] = isnan(fillValue) ? "NaN" : fillValue; 
```
and the opposite when reading the fill-value from json.

Are you interested in making a PR with the necessary changes? Otherwise, I can also do this, but I think I will only have time to look into it in 2 or 3 days.

@meggart 
Ok, so for custom user attributes it seems like this needs to be fixed in `zarr-python`, so that it adheres to the json standard.

**Comment by gdkrmr** [comment]
There is also the case of `Inf` and `-Inf`, I think together wit `NaN`, these are the only "special" values for floats.

@constantinpape I can do it, but I have trouble understanding the test infrastructure, so if you want unit tests with it I would prefer you to do it.

**Comment by gdkrmr** [comment]
Lets do the following: I will try to make it work for my R wrapper and if I get it to work, I make a PR.

**Comment by constantinpape** [comment]
> Lets do the following: I will try to make it work for my R wrapper and if I get it to work, I make a PR.

Sounds good. If it works for you and looks reasonable I can merge it without tests and then write them myself later.

**Comment by constantinpape** [comment]
This was fixed in #149.
Note that the issue with NaN / Inf for user attributes still persists, I will create a separate issue on this.
